### PR TITLE
Pcf max timeout in apply()

### PR DESCRIPTION
**Pull request by emr0327**
- added functionality for a max_timeout in apply() to prevent an infinite loop
- the user must pass the max_timeout integer in seconds as a keyword argument when calling the apply method
- when the max_timeout is reached an exception is raised

### CODE REVIEW

File: pcf/core/particle.py [hunk 1]
```diff
@@ -215,6 +219,9 @@ def apply(self, sync=True, cascade=False, validate_config=False, src_cascade=Non
 
             while not self.is_state_equivalent(self.get_state(), self.desired_state):
                 if self.state == State.pending:
+                    if max_timeout and (time.time() - start_timeout) >= max_timeout:
```
Comment (anovis) [inline pcf/core/particle.py]: i think we only need this check once per while loop and should be done before any logic happens.

**Review by anovis — CHANGES_REQUESTED** [review]
we should add support for these features for quasiparticles as well. this can be done by allowing this to be configured in the pcf_field which then calls each particles apply function.

**Review by ethanwlo — CHANGES_REQUESTED** [review]
Need to add unit tests.

**Comment by emr0327** [comment]
@anovis @bb1314 please review revisions

**Comment by anovis** [comment]
can you add this for quasiparticles as well. it should just be a passthrough in the quasiparticle apply.

**Comment by emr0327** [comment]
sure thing!

File: pcf/test/core/test_quasiparticle.py [hunk 2]
```diff
@@ -177,3 +178,25 @@ def test_rollback():
...
+        "pcf_name":"quasiparticle",
+        "particles":[test_particle_definition],
+        "flavor": "rollback_flavor"
+    }
+    test_quasiparticle = Quasiparticle(test_quasiparticle)
+    test_quasiparticle.set_desired_state(State.running)
+    test_quasiparticle.apply()
+    test_quasiparticle.set_desired_state(State.running)
+    try:
+        test_quasiparticle.apply(max_timeout=2)g
```
Comment (anovis) [inline pcf/test/core/test_quasiparticle.py]: typo. "g" at the end of the line.

**Review by anovis — CHANGES_REQUESTED** [review]
hey the tests are failing for me. did you run with pytest ?

**Comment by anovis** [comment]
the tests still fail. looks like it just goes straight to the assert False statement and doesnt catch the max timeout exception.

**Comment by anovis** [comment]
i think your problem is that you are using the `ParticlePassingVars` class which is from a diferent test suite. that automatically sets the state to running so it skips the while loop where the max timeout is happening. i would choose a different class or create your own for the test. @emr0327

File: pcf/test/core/test_quasiparticle.py [hunk 3]
```diff
@@ -171,9 +171,55 @@ def test_rollback():
...
     quasiparticle_rollback.set_desired_state(State.running)
-    quasiparticle_rollback.apply(rollback=False)
+    with raises(ValueError, message='error'):
+        quasiparticle_rollback.apply(rollback=False)
     # TODO issue with pytest here .fails every other time.
     # assert(quasiparticle_rollback.get_state() == State.pending)
     # assert(quasiparticle_rollback.get_particle("rollback_flavor","pcf_particle_name").state == State.running)
     # assert(quasiparticle_rollback.get_particle("rollback_flavor","particle_error").state == State.terminated)
 
+class QuasiParticleTest(Particle):
```
Comment (anovis) [inline pcf/test/core/test_quasiparticle.py]: can you change this to `TestMaxTimeout`

File: pcf/test/core/test_quasiparticle.py [hunk 4]
```diff
@@ -171,9 +171,55 @@ def test_rollback():
...
+    flavor = "quasiparticle_flavor"
```
Comment (anovis) [inline pcf/test/core/test_quasiparticle.py]: flavor should be quasiparticle

**Review by anovis — CHANGES_REQUESTED** [review]
awesome looks good. these small changes should fix it.

**Comment by emr0327** [comment]
@anovis please review requested changes

**Review by anovis — APPROVED** [review]

**Pull request merged**